Modules/fact_manager.py

```diff
@@ -0,0 +1,336 @@
+"""
+fact_manager.py - Fact Management
+
+Provides a context manager for manipulating, retrieving, and storing facts.
+This class requires database.DatabaseManager and Modules.fact.
+
+Copyright (c) 2018 The Fuel Rat Mischief,
+All rights reserved.
+
+Licensed under the BSD 3-Clause License.
+
+See LICENSE.md
+"""
+import datetime
+import logging
+from datetime import timezone
+
+import psycopg2
+from psycopg2 import sql
+
+
+from Modules.fact import Fact
+from database import DatabaseManager
+
+log = logging.getLogger(f"mecha.{__name__}")
+
+
+class FactManager(DatabaseManager):
+
+    def __init__(self, fact_table="fact2", fact_log="fact_transaction"):
+        if not isinstance(fact_table, str):
+            raise TypeError("fact table must be a string.")
+
+        if not isinstance(fact_log, str):
+            raise TypeError("fact log must be a string.")
+
+        self._FACT_TABLE = fact_table
+        self._FACT_LOG = fact_log
+
+        # Proclaim loudly into the void that we are loaded.
+        log.info("Fact Manager Initialized.")
+        super().__init__()
+
+    async def add(self, fact: Fact):
+        """
+        Adds a new fact to the database.  This will result in a ProgrammingError being thrown
+        if a fact violates the (name, lang) Primary Key relationship.
+
+        VERIFY IF THE FACT EXISTS WITH self.EXISTS() FIRST.
```

Comment:
I wouldn't leave it up to the caller to determine if a fact exists. There's a warning in the comments but nothing code-wise preventing a duplicate from being entered other than an exception. Perhaps raise an error and return to the user cleanly?

Reply:
You're not wrong - but remember - this is the Fact Manager - not the command side.  on the command side (ie user entry side) you'd want to run the .exists check first, and then continue with execution, delivering the message to the user at that point.  I will update the documentation to reflect this.

Reply:
what actually happens when someone tries to add an existing fact? does it explode?

Reply:
If you attempt to insert a fact that already exists, without checking first, the primary key on the table is violated, and you will throw a ``psycopg2.IntegrityError``.  This error is specifically captured and handled in the add method.,